src/api/endpoints/auth/TokenManager.ts

```diff
@@ -37,17 +38,22 @@ export default class TokenManager extends ApiEndpoint {
     // if the refresh token isn't present, nothing we can do
     if (this.refreshToken === undefined) return this.accessToken;
 
-    // token is expired (or will expire soon), so fetch a new one
-    const tokens = await this.fetchToken(this.refreshToken);
+    // token is expired (or will expire soon) and there is not already one fetching, so fetch a new one
+    if (!this.tokenFetchPromise) this.tokenFetchPromise = this.fetchToken(this.refreshToken);
 
-    if (this.refreshCallback !== undefined) {
-      await this.refreshCallback(tokens);
-    }
+    try {
+      const tokens = await this.tokenFetchPromise;
+      this.accessToken = tokens.accessToken;
+      this.refreshToken = tokens.refreshToken;
 
-    this.accessToken = tokens.accessToken;
-    this.refreshToken = tokens.refreshToken;
+      if (this.refreshCallback !== undefined) {
+        await this.refreshCallback(tokens);
+      }
```

Comment:
we still risk calling refreshCallback multiple times - is that fine?

Reply:
Yes because the tokens will be the exact same